lib/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.cpp

```diff
@@ -466,6 +467,72 @@ struct GatherScatterOpPattern : public OpConversionPattern<OpT> {
   }
 };
 
+// Given a tensor and its representation in tensor memory, determine its
+// distributed layout.
+static RankedTensorType getTMEMTensorLayout(const TypeConverter *tc,
+                                            RankedTensorType type,
+                                            MemDescType memdesc,
+                                            unsigned numWarps) {
+  Attribute encoding;
+  type = cast<RankedTensorType>(tc->convertType(type));
+  if (isa<ttng::TensorMemoryScalesEncodingAttr>(memdesc.getEncoding())) {
+    encoding = LinearEncodingAttr::get(
+        type.getContext(), getScaleTMEMStoreLinearLayout(type, numWarps));
+  } else {
+    auto tmemEnc = cast<ttng::TensorMemoryEncodingAttr>(memdesc.getEncoding());
+    encoding = ttng::getTmemCompatibleLayout(
+        tmemEnc.getBlockM(), tmemEnc.getBlockN(), type, numWarps);
+  }
+  return RankedTensorType::get(type.getShape(), type.getElementType(),
+                               encoding);
+}
+
+struct TMEMLoadOpPattern : public OpConversionPattern<ttng::TMEMLoadOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ttng::TMEMLoadOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    RankedTensorType type = getTMEMTensorLayout(
+        typeConverter, op.getType(), op.getSrc().getType(), lookupNumWarps(op));
+    rewriter.modifyOpInPlace(op, [&] { op.getResult().setType(type); });
+    return success();
+  }
+};
+
+struct TMEMStoreOpPattern : public OpConversionPattern<ttng::TMEMStoreOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ttng::TMEMStoreOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    RankedTensorType type =
+        getTMEMTensorLayout(typeConverter, op.getSrc().getType(),
+                            op.getDst().getType(), lookupNumWarps(op));
+    Value src =
+        rewriter.create<ConvertLayoutOp>(op.getLoc(), type, adaptor.getSrc());
+    rewriter.modifyOpInPlace(op, [&] { op.getSrcMutable().assign(src); });
+    return success();
+  }
+};
```

Comment:
nit: feels a bit weird to have those patterns in TritonToTritonGPUPass. I wonder if we can find a better layering? Doesn't have to be addressed right now

Reply:
Yes agreed. I think actually the better way to do it is to have a separate pass entirely and we can run that before or after convert-triton-to-tritongpu during relayout.

Reply:
yeah that would be better, feels ugly to have triton to triton-gpu handle triton-gpu inputs :) 
I'm fine if you prefer to do it in a follow up

Reply:
I'll refactor this in a follow up